src/mpid/ch4/netmod/ofi/ofi_events.c

```diff
@@ -652,8 +652,6 @@ static int am_recv_event(struct fi_cq_tagged_entry *wc, MPIR_Request * rreq)
             break;
 
         case MPIDI_AMTYPE_RDMA_READ:
-            /* buffer always copied together (there is no payload, just LMT header) */
-            p_data = (char *) am_hdr + sizeof(*am_hdr) + am_hdr->am_hdr_sz;
             mpi_errno = MPIDI_OFI_handle_rdma_read(am_hdr, am_hdr + 1,
                                                    (MPIDI_OFI_lmt_msg_payload_t *) p_data);
```

Comment:
What is the actual bug in the original code?  Is it the p_data assignment in `MPIDI_AMTYPE_RDMA_READ`?

Reply:
yes, in `MPIDI_AMTYPE_RDMA_READ`, p_data assignment is wrong, it should use `orig_buf`

Reply:
The commit changed more than what needs to be fixed. That is why I was not sure what was the actual bug. Can you change only the line under `MPIDI_AMTYPE_RDMA_READ`?

Reply:
I read the code again to figure out the diff of `am_hdr` and `orig_buf`. Both of them are initially set to `wc->buf` at line 592. But `am_hdr` is updated to a temp buffer if `NEEDS_STRICT_ALIGNMENT` is on. I guess this is the root cause of the bug. In other words, if `NEEDS_STRICT_ALIGNMENT` is off, either `am_hdr` or `orig_buf` is fine as they are identical. Can you please explain it explicitly in the commit message? Your commit message does not tell me such info.

Reply:
BTW, not only does the bug exist under `MPIDI_AMTYPE_RDMA_READ`, it also exists under `MPIDI_AMTYPE_PIPELINE`, `wc->buf` should be changed to `orig_buf` too. So I would change them both.

Reply:
Why should `wc->buf` be `orig_buf `? I see they are identical. I.e., `void *orig_buf = wc->buf;  ` at line 592.

Reply:
Yes, the issue of using `wc->buf` is when the packet is not next seq packet, it will dequeue previous cached packet through `MPIDI_OFI_am_claim_unordered_msg` at line 674. The `orig_buf` is updated here, but `wc->buf` is the wrong buffer pointer

Reply:
I see. Good catch. The bug is not obvious. Can you add explanation for both cases in the commit message?

Reply:
Moreover, since `MPIDI_AMTYPE_SHORT_HDR` does not use `p_data`. Let us just change the two lines under `MPIDI_AMTYPE_RDMA_READ` and `MPIDI_AMTYPE_PIPELINE ` to minimize the code change

Reply:
ok, will do